presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -1363,6 +1375,18 @@ private ThriftMetastoreClient createMetastoreClient()
         return clientProvider.createMetastoreClient();
     }
 
+    private ThriftMetastoreClient createMetastoreClient(HiveContext context)
+            throws TException
+    {
+        ThriftMetastoreClient client = createMetastoreClient();
+        if (!isMetastoreImpersonationEnabled || hiveMetastoreAuthentication instanceof NoHiveMetastoreAuthentication) {
+            return client;
+        }
+
+        client.setUGI(context.getUsername());
+        return clientProvider.createMetastoreClient();
```

Comment:
Shouldn't we be returning `client` here? Creating a new client here after setting the UGI seems wrong.